File: python3/max_consecutive_ones_iii.py
# Space : O(1)
# Time  : O(n)

class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        ones, zeros, start = 0, 0, 0
        n, ans = len(nums), 0

        for i in range(n):
            if nums[i] == 0:
                zeros += 1
            else:
                ones += 1

            while zeros > k:
                if nums[start] == 1:
                    ones -= 1
                else:
                    zeros -= 1
                start += 1

            ans = max(ans, i - start + 1)

        return ans

File: golang/max_consecutive_ones_iii.go
package main

// Space : O(n)
// Time  : O(n)

func longestOnes(nums []int, k int) int {
    ones, zeros, start, ans := 0, 0, 0, 0
    n := len(nums)
    
    for i := 0; i < n; i++ {
        if nums[i] == 1 {
            ones++
        } else {
            zeros++
        }
        
        for zeros > k {
            if nums[start] == 1 {
                ones--
            } else {
                zeros--
            }
            start++
        }
        
        ans = Max(ans, i-start+1)
    }
    
    return ans
}
